src/commands/apphosting-rollouts-create.ts

```diff
@@ -1,27 +1,116 @@
+import { getRepoDetailsFromBackend, listAllBranches } from "../gcp/devConnect";
+import {
+  getGitHubBranch,
+  getGitHubCommit,
+  promptGitHubBranch,
+} from "../apphosting/githubConnections";
+import { orchestrateRollout } from "../apphosting";
+import { confirm } from "../prompt";
+import { logBullet } from "../utils";
+import { consoleOrigin } from "../api";
 
-export const command = new Command("apphosting:rollouts:create <backendId> <buildId>")
+export const command = new Command("apphosting:rollouts:create <backendId>")
   .description("create a rollout using a build for an App Hosting backend")
   .option("-l, --location <location>", "specify the region of the backend", "us-central1")
   .option("-i, --id <rolloutId>", "id of the rollout (defaults to autogenerating a random id)", "")
+  .option("-b, --git-branch <gitBranch>", "repository branch to deploy (defaults to 'main')", "")
+  .option("-c, --commit <commit>", "git commit to deploy (defaults to the latest commit)", "")
+  .option("-f, --force", "skip confirmation")
   .before(apphosting.ensureApiEnabled)
-  .action(async (backendId: string, buildId: string, options: Options) => {
+  .action(async (backendId: string, options: Options) => {
     const projectId = needProjectId(options);
     const location = options.location as string;
-    // TODO: Should we just reuse the buildId?
-    const rolloutId =
-      (options.buildId as string) ||
-      (await apphosting.getNextRolloutId(projectId, location, backendId));
-    const build = `projects/${projectId}/backends/${backendId}/builds/${buildId}`;
-    const op = await apphosting.createRollout(projectId, location, backendId, rolloutId, {
-      build,
+
+    let branch = options.gitBranch as string | undefined;
+    const commit = options.commit as string | undefined;
+    if (branch && commit) {
+      throw new FirebaseError(
+        "Cannot specify both a branch and commit to deploy. Please specify either --git-branch or --commit.",
+      );
+    }
+    const backend = await apphosting.getBackend(projectId, location, backendId);
+    if (!backend.codebase.repository) {
+      throw new FirebaseError(
+        `Failed to find connected Git repository for backend ${backendId}. Please check that your backend has been successfully linked to a Git repository.`,
+      );
+    }
+    const { repoLink, owner, repo, readToken } = await getRepoDetailsFromBackend(
+      projectId,
+      location,
+      backend.codebase.repository,
+    );
+
+    let target;
+    if (branch) {
+      const branches = await listAllBranches(repoLink.name);
+      if (!branches.has(branch)) {
+        throw new FirebaseError(
+          `Unrecognized git branch ${branch}. Please double-check your branch name and try again.`,
+        );
+      }
+      const branchInfo = await getGitHubBranch(owner, repo, branch, readToken.token);
+      target = branchInfo.commit;
+    } else if (commit) {
+      if (!/^(?:[0-9a-f]{40}|[0-9a-f]{7})$/.test(commit)) {
```

Comment:
consider pulling this up to a file level const with a proper name to make it clear that this is our commit sha regex